File: java/example_code/ec2/src/main/java/aws/example/ec2/AllocateAddress.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package aws.example.ec2;

import com.amazonaws.services.ec2.AmazonEC2;
import com.amazonaws.services.ec2.AmazonEC2ClientBuilder;
import com.amazonaws.services.ec2.model.AllocateAddressRequest;
import com.amazonaws.services.ec2.model.AllocateAddressResult;
import com.amazonaws.services.ec2.model.AssociateAddressRequest;
import com.amazonaws.services.ec2.model.AssociateAddressResult;
import com.amazonaws.services.ec2.model.DomainType;

/**
 * Allocates an elastic IP address for an EC2 instance
 */
public class AllocateAddress {
        public static void main(String[] args) {
                final String USAGE = "To run this example, supply an instance id\n" +
                                "Ex: AllocateAddress <instance_id>\n";

                if (args.length != 1) {
                        System.out.println(USAGE);
                        System.exit(1);
                }

                String instance_id = args[0];

                final AmazonEC2 ec2 = AmazonEC2ClientBuilder.defaultClient();

                AllocateAddressRequest allocate_request = new AllocateAddressRequest()
                                .withDomain(DomainType.Vpc);

                AllocateAddressResult allocate_response = ec2.allocateAddress(allocate_request);

                String allocation_id = allocate_response.getAllocationId();

                AssociateAddressRequest associate_request = new AssociateAddressRequest()
                                .withInstanceId(instance_id)
                                .withAllocationId(allocation_id);

                AssociateAddressResult associate_response = ec2.associateAddress(associate_request);

                System.out.printf(
                                "Successfully associated Elastic IP address %s " +
                                                "with instance %s",
                                associate_response.getAssociationId(),
                                instance_id);
        }
}

File: go/ec2/AllocateAddress/AllocateAddress.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// snippet-start:[ec2.go.allocate_address]
package main

// snippet-start:[ec2.go.allocate_address.import]
import (
    "flag"
    "fmt"

    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/ec2"
    "github.com/aws/aws-sdk-go/service/ec2/ec2iface"
)
// snippet-end:[ec2.go.allocate_address.import]

// AllocateAndAssociate allocates a VPC Elastic IP address with an instance.
// Inputs:
//     svc is an Amazon EC2 service client
//     instanceID is the ID of the instance
// Output:
//     If success, information about the allocation and association and nil
//     Otherwise, two nils and an error from the call to AllocateAddress or AssociateAddress
func AllocateAndAssociate(svc ec2iface.EC2API, instanceID *string) (*ec2.AllocateAddressOutput, *ec2.AssociateAddressOutput, error) {
    // snippet-start:[ec2.go.allocate_address.allocate]
    allocRes, err := svc.AllocateAddress(&ec2.AllocateAddressInput{
        Domain: aws.String("vpc"),
    })
    // snippet-end:[ec2.go.allocate_address.allocate]
    if err != nil {
        return nil, nil, err
    }

    // snippet-start:[ec2.go.allocate_address.associate]
    assocRes, err := svc.AssociateAddress(&ec2.AssociateAddressInput{
        AllocationId: allocRes.AllocationId,
        InstanceId:   instanceID,
    })
    // snippet-end:[ec2.go.allocate_address.associate]
    if err != nil {
        return nil, nil, err
    }

    return allocRes, assocRes, nil
}

func main() {
    // snippet-start:[ec2.go.allocate_address.args]
    instanceID := flag.String("i", "", "The ID of the instance")
    flag.Parse()

    if *instanceID == "" {
        fmt.Println("You must supply the ID of an instance")
        return
    }
    // snippet-end:[ec2.go.allocate_address.args]

    // snippet-start:[ec2.go.allocate_address.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    svc := ec2.New(sess)
    // snippet-end:[ec2.go.allocate_address.session]

    allocRes, assocRes, err := AllocateAndAssociate(svc, instanceID)
    if err != nil {
        fmt.Println("Got an error attempting to associate IP address to instance with ID "+*instanceID, ":")
        fmt.Println(err)
        return
    }

    // snippet-start:[ec2.go.allocate_address.display]
    fmt.Println("Allocated IP address", *allocRes.PublicIp, "with instance", *instanceID)
    fmt.Println("Allocation id:", *allocRes.AllocationId, "association id: ", *assocRes.AssociationId)
    // snippet-end:[ec2.go.allocate_address.display]
}
// snippet-end:[ec2.go.allocate_address]
